File: 0503. Next Greater Element II/stack.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        // index
        Stack<Integer> monoStack = new Stack<>();
        Set<Integer> visited = new HashSet<>();
        int len = nums.length;
        int[] ret = new int[len];
        int[] v = new int[2 * len - 1];
        Arrays.fill(ret, -1);
        for(int i = 0; i < v.length; ++i){
            v[i] = nums[i % len];
        }

        for(int i = 0; i < v.length; ++i){
            int cur = v[i];
            while(!monoStack.empty() && v[monoStack.peek() % len] < cur){
                if(!visited.contains(monoStack.peek() % len)){
                    int index = monoStack.pop();
                    ret[index] = cur;
                    visited.add(index);
                }else{
                    monoStack.pop();
                }
            }
            monoStack.push(i);
        }
        
        return ret;
    }
}

File: cn/0735. 行星碰撞/stack.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> pos;
        for(const int& i: asteroids) {
            if(pos.empty()){
                pos.push(i);
                continue;
            }
            
            bool isDestroy = false;
            if(i < 0){
                while(!pos.empty()){
                    int top = pos.top();
                    if(top < 0){
                        break;
                    }
                    if(top < -i){
                        pos.pop();
                    }else if(top == -i){
                        isDestroy = true;
                        pos.pop();
                        break;
                    }else{
                        isDestroy = true;
                        break;
                    }
                }
            }
            if(!isDestroy){
                pos.push(i);
            }
        }

        vector<int> ret;
        while(!pos.empty()){
            ret.push_back(pos.top());
            pos.pop();
        }
        reverse(ret.begin(), ret.end());
        return ret;
    }
};